TitlePart with RenderTitle option and Placement for Form, Label

ns8482e opened the pull request:
Fixes #6079 
Fixes #5959 

/cc @deanmarcussen @Piedone  @jwu-au

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Forms/placement.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "TitlePart": [
```
Why is is necessary since this can be done with the settings now?

ns8482e replied:
To always hide? - regardless of setting. May be we can remove placement  entry for Form but not for Label

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Title/Drivers/TitlePartDisplay.cs:
```diff
@@ -21,13 +21,20 @@ public TitlePartDisplay(IStringLocalizer<TitlePartDisplay> localizer)
+                   model.Title = titlePart.ContentItem.DisplayText;
+                   model.TitlePart = titlePart;
+               })
+               .Location("Detail", "Header:5")
```
What is the display type that is used in the Container case in the admin? If it's Detail we might have an issue with the ComonMetadata shape that renders a <title> value with display text. (unrelated).
Should we have different settings for the front-end and the backend (not sure we are even using distinct display types)?

ns8482e replied:
I didn't change Container def in this PR.  Container doesn't have Title yet.

ns8482e replied:
@sebastienros RenderTitle is for front-end. For Admin "Edit" shape Editable and EditableRequired is used

deanmarcussen replied:
>  What is the display type that is used in the Container case in the admin? If it's Detail we might have an issue with the ComonMetadata shape that renders a <title> value with display text. (unrelated).

For Widgets (Detail) we have a `Widget-ContentsMetadata.cshtml` (in the flow module) which provides an empty shape.

For BagParts (in Detail) it remains an unrelated issue

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Title/Models/TitlePartSettings.cs:
```diff
@@ -22,5 +22,10 @@ public class TitlePartSettings
+        /// <summary>
+        /// Render TitlePart.
+        /// </summary>
+        public bool RenderTitle { get; set; }
```
Should be true by default for backward compatibilty.

ns8482e replied:
I guess default should be what we want to keep going forward. Is it possible to use migration here for backward compatibility?

ns8482e replied:
I tried with `public bool RenderTitle { get; set; } = true;` but migration is not working

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Title/Views/TitlePartSettings.Edit.cshtml:
```diff
@@ -21,6 +21,14 @@
+    <div class="custom-control custom-checkbox">
+        <input asp-for="RenderTitle" class="custom-control-input" />
+        <label asp-for="RenderTitle" class="custom-control-label">@T["Render Title"]</label>
+        <span class="hint">@T["Check to renter Title. <br/> (It is suggested to use ContentItem.DisplayText in template and keep this unchecked.)"]</span>
```
```suggestion
        <span class="hint">@T["Check to renter the title in views.)"]</span>
```

I don't want to mention to use ContentItem.DisplayText as we could also use this PR to remove the Title property off the part. Or mark it as deprecated at least. It shouldn't be stored or used.

Piedone replied:
There's a typo and also I'd use something more descriptive instead of "views" like "Check to render the title on the frontend."

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Title/Views/TitlePartSettings.Edit.cshtml:
```diff
@@ -21,6 +21,14 @@
+    <div class="custom-control custom-checkbox">
+        <input asp-for="RenderTitle" class="custom-control-input" />
+        <label asp-for="RenderTitle" class="custom-control-label">@T["Render Title"]</label>
+        <span class="hint">@T["Check to renter the title in views.)"]</span>
```
```suggestion
        <span class="hint">@T["Check to render the title in views.)"]</span>
```

Did you disable the permssion for us to update your PRs directly?

ns8482e replied:
No I didn’t change anything

Piedone commented on src/OrchardCore.Modules/OrchardCore.Title/Models/TitlePartSettings.cs:
```diff
@@ -22,5 +22,10 @@ public class TitlePartSettings
         public string Pattern { get; set; }
+
+        /// <summary>
+        /// Render TitlePart.
```
Use something that describes what this is for instead of repeating the property name. Imagine you know nothing about this feature and see the property named `RenderTitle`. What does it mean? What's "render"? Where does this rendering happen? You can also rename the property to make it more descriptive instead of relying the comment.

ns8482e commented:
So, Now, I changed the default `RenderTitle` to `true`. Also deleted `placement.json` and added migration for Form and Label. However there is little problem.

1. To use  `MergeSettings<T>( Action<T> )` it requires me to add direct project reference to `OrchardCore.Title`. IMO it `MergeSettings` should have non typed version that should take setting name and anonymous object.

2. `MergeSettings `ignores default datatype value. Here `false` is default for `bool `hence it's never gets written to TitlePartSettings but now, default RenderTitle is to true in C#. It always display as true.

deanmarcussen commented:
> So, Now, I changed the default `RenderTitle` to `true`. Also deleted `placement.json` and added migration for Form and Label. However there is little problem.
> 
> 1. To use  `MergeSettings<T>( Action<T> )` it requires me to add direct project reference to `OrchardCore.Title`. IMO  `MergeSettings` should have non typed version that should take setting name and anonymous object to avoid adding project reference.

What we do here is just create an internal class with the same settings name. This example (edited) from the taxonomies migrations

```
    internal class AutoroutePartSettings
    {
        public string Pattern { get; set; }

        [DefaultValue(true)]   // you need this
        public bool AllowRouteContainedItems { get; set; } = true; // and this
    }
```


> 2. `MergeSettings `ignores default datatype value while saving. Here `false` is default for `bool `hence it's never gets written to TitlePartSettings but now, default RenderTitle is set to true in C#. So It always display as true even after running the migration with `false `value.

If you use `[DefaultValue(true)]` will work with the json serializer.

ns8482e commented:
> What we do here is just create an internal class with the same settings name

It works but requires to define internal class exactly same as original model class. When I defined only concerned property that i am using in migration as below it ignores other properties.

```
internal class TitlePartSettings
{
      [DefaultValue(true)]
      public bool RenderTitle { get; set; }
}
```

So, I have to define as below.

```
internal class TitlePartSettings
{
      [DefaultValue(TitlePartOptions.Editable)]
      public TitlePartOptions Options { get; set; }

      public string Pattern { get; set; }

      [DefaultValue(true)]
      public bool RenderTitle { get; set; }
}
```

I still feel it would be better to use anonymous class and use JsonMerge.  https://www.newtonsoft.com/json/help/html/MergeJson.htm

As name of settings is always be `[PartName]Setting`

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Forms/Migrations.cs:
```diff
@@ -100,7 +105,29 @@ public int Create()
+
+            return 2;
+        }
+        internal class TitlePartSettings
```
can you put a new line here

ns8482e replied:
thanks- done

Piedone commented on src/OrchardCore.Modules/OrchardCore.Title/Models/TitlePartSettings.cs:
```diff
@@ -22,5 +22,11 @@ public class TitlePartSettings
         public string Pattern { get; set; }
+
+        /// <summary>
+        /// Get or Set whether to render the title in views
```
```suggestion
        /// Gets or sets whether to render the title in display views.
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Title/Views/TitlePartSettings.Edit.cshtml:
```diff
@@ -21,6 +21,14 @@
+    <div class="custom-control custom-checkbox">
+        <input asp-for="RenderTitle" class="custom-control-input" />
+        <label asp-for="RenderTitle" class="custom-control-label">@T["Render Title"]</label>
+        <span class="hint">@T["Check to render the title in views.)"]</span>
```
```suggestion
        <span class="hint">@T["Check to render the title in display views."]</span>
```

ns8482e commented:
@deanmarcussen @sebastienros It would be nice if this is available in RC2

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Title/Views/TitlePartSettings.Edit.cshtml:
```diff
@@ -21,6 +21,14 @@
+    <div class="custom-control custom-checkbox">
+        <input asp-for="RenderTitle" class="custom-control-input" />
+        <label asp-for="RenderTitle" class="custom-control-label">@T["Render Title"]</label>
+        <span class="hint">@T["Check to render the title in display views."]</span>
```
```suggestion
        <span class="hint">@T["Whether to render the title in display views."]</span>
```
